### PR TITLE
JAMES-3058 [Part 2] Expose task solving mailbox inconsistencies over webadmin

### PR DESCRIPTION
**Pull request by chibenwa**

### CODE REVIEW

File: server/container/guice/cassandra-guice/src/main/java/org/apache/james/modules/webadmin/InconsistencySolvingRoutesModule.java [hunk 1]
```diff
@@ -17,7 +17,7 @@
  * under the License.                                           *
  ****************************************************************/
 
-package org.apache.james.modules.server;
+package org.apache.james.modules.webadmin;
```
Comment (mbaechler) [inline server/container/guice/cassandra-guice/src/main/java/org/apache/james/modules/webadmin/InconsistencySolvingRoutesModule.java]: why did you put webadmin stuff in a guice project?

File: src/site/markdown/server/manage-webadmin.md [hunk 2]
```diff
@@ -358,7 +358,45 @@ Response codes:
...
-### Recomputing JMAP fast message view projection
+#### Fixing mailboxes inconsistencies
+
+This task is only available on top of Guice Cassandra products.
+
+```
+curl -XPOST /mailboxes?task=solveInconsistencies
+```
+
+Will schedule a task for fixing the mailbox projection.
```
Comment (mbaechler) [inline src/site/markdown/server/manage-webadmin.md]: what is "mailbox projection"?

File: src/site/markdown/server/manage-webadmin.md [hunk 3]
```diff
@@ -358,7 +358,45 @@ Response codes:
...
+
+[More details about endpoints returning a task](#Endpoints_returning_a_task).
+
+The scheduled task will have the following type `solve-mailbox-inconsistencies` and the following `additionalInformation`:
+
+```
+{
+  "type":"solve-mailbox-inconsistencies",
+  "processedMailboxEntries": 3,
```
Comment (mbaechler) [inline src/site/markdown/server/manage-webadmin.md]: you don't report on fixed things?

File: server/container/guice/cassandra-guice/src/main/java/org/apache/james/modules/webadmin/InconsistencySolvingRoutesModule.java (hunk 1 above)
Comment (chibenwa) [reply server/container/guice/cassandra-guice/src/main/java/org/apache/james/modules/webadmin/InconsistencySolvingRoutesModule.java]: I did put some guice registrations in a guice project. What's wrong?

File: src/site/markdown/server/manage-webadmin.md (hunk 2 above)
Comment (chibenwa) [reply src/site/markdown/server/manage-webadmin.md]: s/mailbox projection/mailbox deduplicated object/ ?

File: src/site/markdown/server/manage-webadmin.md (hunk 3 above)
Comment (chibenwa) [reply src/site/markdown/server/manage-webadmin.md]: That what `fixedInconsistencies` is about. Processed intent  is showing progress.

Or do you want a list of fixed ids?

File: src/site/markdown/server/manage-webadmin.md [hunk 4]
```diff
@@ -435,6 +435,9 @@ Response codes:
 
 #### ReIndexing action
 
+These tasks is only available on top of Guice Cassandra products or Guice JPA products. They are not part of Memory
```
Comment (Arsnael) [inline src/site/markdown/server/manage-webadmin.md]: s/is/are

**Comment by Arsnael** [comment]
Needs a rebase

**Comment by chibenwa** [comment]
test this please

(build was aborted)

File: src/site/markdown/server/manage-webadmin.md [hunk 5]
```diff
@@ -364,7 +364,45 @@ Response codes:
...
+    "mailboxPathDaoEntry":{
+      "mailboxPath":"#private:user:mailboxName2",
+      "mailboxId":"464765a0-e4e7-11e4-aba4-710c1de3782b"
+    }
+  }]
+}
+```
+
+Note that conflicting entry inconsistencies will not be fixed and will require to explicitly use 
+[ghost mailbox](#correcting-ghost-mailbox) endpoint in order to merge the conflicting mailboxes and prevent any message
```
Comment (trantienduchn) [inline src/site/markdown/server/manage-webadmin.md]: careful the '-' doesn't work on apache site, you should use '_'

**Comment by chibenwa** [comment]
test this please

File: server/container/guice/cassandra-guice/src/main/java/org/apache/james/modules/webadmin/InconsistencySolvingRoutesModule.java (hunk 1 above)
Comment (mbaechler) [reply server/container/guice/cassandra-guice/src/main/java/org/apache/james/modules/webadmin/InconsistencySolvingRoutesModule.java]: That's ok

File: src/site/markdown/server/manage-webadmin.md (hunk 3 above)
Comment (mbaechler) [reply src/site/markdown/server/manage-webadmin.md]: The use case is :  
1. bob asks his admin why there's a mailbox missing (or whatever inconsistencies is about)
2. admin run the solve-mailbox-inconsistencies task in the hope of fixing it
3. admin don't have the information in the report to know if a mailbox for this user has been fixed

Comment (chibenwa) [reply src/site/markdown/server/manage-webadmin.md]: So we want the list of fixed ids ;-)

**Comment by chibenwa** [comment]
Force pushed to solve conflict.

@rouazana  only review of the two last commits is needed.

**Comment by Arsnael** [comment]
Seems that build was running on the slave that crashed yesterday.... No result.

test this please

**Comment by chibenwa** [comment]
```
[d1923f8478891a06501451eabafd375b11cc5fcd] [ERROR]   RabbitMQWebAdminServerIntegrationTest.solveMailboxInconsistenciesTaskShouldBeExposed:180 1 expectation failed.
[d1923f8478891a06501451eabafd375b11cc5fcd] JSON path status doesn't match.
```

**Comment by chibenwa** [comment]
```
[d1923f8478891a06501451eabafd375b11cc5fcd] [ERROR]   RabbitMQWebAdminServerIntegrationTest.solveMailboxInconsistenciesTaskShouldBeExposed:180 1 expectation failed.
[d1923f8478891a06501451eabafd375b11cc5fcd] JSON path status doesn't match.
```

**Comment by Arsnael** [comment]
Logs not found anymore...

test this please

**Comment by Arsnael** [comment]
Logs not found anymore...

test this please

**Comment by Arsnael** [comment]
```
[3dcef285f185c636865ccbe7d502d3a00b6ca85f] [ERROR] Failures: 
[3dcef285f185c636865ccbe7d502d3a00b6ca85f] [ERROR]   RabbitMQWebAdminServerIntegrationTest.solveMailboxInconsistenciesTaskShouldBeExposed:181 1 expectation failed.
[3dcef285f185c636865ccbe7d502d3a00b6ca85f] JSON path status doesn't match.
[3dcef285f185c636865ccbe7d502d3a00b6ca85f] Expected: is "completed"
[3dcef285f185c636865ccbe7d502d3a00b6ca85f]   Actual: failed
```
test this please

**Comment by trantienduchn** [comment]
> ```
> [3dcef285f185c636865ccbe7d502d3a00b6ca85f] [ERROR] Failures: 
> [3dcef285f185c636865ccbe7d502d3a00b6ca85f] [ERROR]   RabbitMQWebAdminServerIntegrationTest.solveMailboxInconsistenciesTaskShouldBeExposed:181 1 expectation failed.
> [3dcef285f185c636865ccbe7d502d3a00b6ca85f] JSON path status doesn't match.
> [3dcef285f185c636865ccbe7d502d3a00b6ca85f] Expected: is "completed"
> [3dcef285f185c636865ccbe7d502d3a00b6ca85f]   Actual: failed
> ```
> 

the test `solveMailboxInconsistenciesTaskShouldBeExposed` belongs to this PR

**Comment by Arsnael** [comment]
> > ```
> > [3dcef285f185c636865ccbe7d502d3a00b6ca85f] [ERROR] Failures: 
> > [3dcef285f185c636865ccbe7d502d3a00b6ca85f] [ERROR]   RabbitMQWebAdminServerIntegrationTest.solveMailboxInconsistenciesTaskShouldBeExposed:181 1 expectation failed.
> > [3dcef285f185c636865ccbe7d502d3a00b6ca85f] JSON path status doesn't match.
> > [3dcef285f185c636865ccbe7d502d3a00b6ca85f] Expected: is "completed"
> > [3dcef285f185c636865ccbe7d502d3a00b6ca85f]   Actual: failed
> > ```
> 
> the test `solveMailboxInconsistenciesTaskShouldBeExposed` belongs to this PR

Could observe something really odd in my local, the task is asserting that schema version is at least 6, but it would fail because the version is empty sometimes when playing the all test suite

So I enforce first an upgrade to the latest schema version at the beginning of the test

**Comment by Arsnael** [comment]
```
[18ccf02ddb61e3029a1ef4db618f6f2bdd3ca325] [ERROR] Errors: 
[18ccf02ddb61e3029a1ef4db618f6f2bdd3ca325] [ERROR]   MessageFullViewFactoryTest$WithProjectionInvolvedTest.fromMessageResultsShouldComputeWhenProjectionReturnEmpty:883 » NullPointer
```
the PR dealing with this has not been merged yet... 

test this please

**Comment by Arsnael** [comment]
```
[18ccf02ddb61e3029a1ef4db618f6f2bdd3ca325] [ERROR] Failures: 
[18ccf02ddb61e3029a1ef4db618f6f2bdd3ca325] [ERROR]   ObjectStorageBlobStoreAWSPrefixTest.saveStringShouldPublishSaveBytesTimerMetrics 
[18ccf02ddb61e3029a1ef4db618f6f2bdd3ca325] Expected size:<2> but was:<1> in:
[18ccf02ddb61e3029a1ef4db618f6f2bdd3ca325] <[PT0.124945548S, PT0.013678503S]>
```
Doesn't seem related...

**Comment by Arsnael** [comment]
test this please

**Comment by Arsnael** [comment]
Merged

**Pull request closed without merging**